File: python/quick_sort.py


def troca(vetor: list, i: int, j: int):
	aux = vetor[i]
	vetor[i] = vetor[j]
	vetor[j] = aux


def particao(vetor: list, inicio: int, fim: int) -> int:
	pivo = vetor[(inicio + fim) // 2]
	i = inicio - 1
	j = fim + 1

	while i < j:
		
		j -= 1
		while vetor[j] > pivo:
			j -= 1

		i += 1
		while vetor[i] < pivo:
			i += 1

		if i < j:
			troca(vetor, i, j)
	return j


def quick_sort(vetor: list, inicio: int, fim: int):
	if inicio < fim:
		q = particao(vetor, inicio, fim)
		quick_sort(vetor, inicio, q)
		quick_sort(vetor, (q + 1), fim)


if __name__ == '__main__':
	
	N = 10
	vetor = [None] * N

	for i in range(N):
		vetor[i] = int(input('Entre com o ' + str(i + 1) + 'º valor: '))

	quick_sort(vetor, 0, (N - 1))

	for i in range(N):
		print(str(i + 1) + '° número: ' + str(vetor[i]), end='\n')

File: java/QuickSort.java

import java.util.Scanner;

class QuickSort {
	public static void main(String[] args) {

		int N = 10;
		int vetor[] = new int[N];
		int aux;
		Scanner entrada = new Scanner(System.in);

		for (int i = 0; i < N; i++) {
			System.out.print("Digite o " + (i + 1) + "° número: ");
			vetor[i] = entrada.nextInt();
		}

		quicksort(vetor, 0, (N - 1));

		for (int i = 0; i < N; i++) {
			System.out.println((i + 1) + "° número: " + vetor[i]);
		}
	}

	private static void quicksort(int vetor[], int inicio, int fim) {
		int q;
		if (inicio < fim) {
			q = particao(vetor, inicio, fim);
			quicksort(vetor, inicio, q);
			quicksort(vetor, (q + 1), fim);
		}
	}

	private static int particao(int vetor[], int inicio, int fim) {
		int pivo = vetor[(inicio + fim) / 2];
		int i = inicio - 1;
		int j = fim + 1;

		while (i < j) {

			do {
				j -= 1;
			} while (vetor[j] > pivo);

			do {
				i += 1;
			} while (vetor[i] < pivo);

			if (i < j) {
				troca(vetor, i, j);
			}
		}
		return j;
	}

	private static void troca(int vetor[], int i, int j) {
		int aux = vetor[i];
		vetor[i] = vetor[j];
		vetor[j] = aux;
	}
}
